Calibration/IsolatedParticles/src/CaloPropagateTrack.cc

```diff
@@ -601,6 +602,27 @@ namespace spr{
     return spr::propagateECAL (vertex, momentum, charge, bfield, debug);
   }
 
+  std::pair<DetId,bool> propagateIdECAL(const HcalDetId& id, const CaloGeometry* geo, const MagneticField* bField, bool debug) {
+
+    const HcalGeometry* gHB = (const HcalGeometry*)(geo->getSubdetectorGeometry(DetId::Hcal,HcalBarrel));
+    GlobalPoint vertex(0,0,0);
+    GlobalPoint hit(gHB->getPosition(id));
+    GlobalVector momentum = GlobalVector(hit.x(),hit.y(),hit.z());
+    std::pair<math::XYZPoint,bool> info = propagateECAL(vertex,momentum,0,bField,debug);
+    DetId eId(0);
+    if (info.second) {
+      const GlobalPoint point(info.first.x(),info.first.y(),info.first.z());
+      if (std::abs(point.eta())<spr::etaBEEcal) {
+	const EcalBarrelGeometry *barrelGeom = (dynamic_cast< const EcalBarrelGeometry *> (geo->getSubdetectorGeometry(DetId::Ecal,EcalBarrel)));
```

Comment:
I don't think this cast is needed. getClosestCell is part of the base class.

Reply:
This is overridden in the EcalBarrelGeometry.h, EcalEndcapGeometry.h, HcalGeometry.H

Reply:
There is no way I can avoid the casting here

Reply:
hi @bsunanda -i'm not sure I follow - your code does not need to know the type - as of course the pointer will know the type and call the appropriate function (unless something strange has been done upstream?)